### PR TITLE
Make multi-instrument idioms for async callbacks possible

### LINKED ISSUE #2280: Enable multi-instrument callbacks in the metrics API
**What are you trying to achieve?**

The way the specification for asynchronous instrument callbacks is written leaves a lot of flexibility to the SDK. This issue requests a little bit more flexibility to support multiple instruments per callback. Discussion in #2226 and #2249 leads to the conclusion that multiple callbacks per instrument should be allowed (see proposed wording #2270). This issue is about multiple instruments per callback.

In OTel-Go, a common example for asynchronous instruments is to report memory statistics. The [runtime.ReadMemStats](https://pkg.go.dev/runtime#ReadMemStats) API captures a bunch of associated measurements, which translate into multiple distinct asynchronous metrics, so the natural pattern is to let one callback support multiple asynchronous instruments.

One API proposal for asynchronous instruments in OTel-Go would make them appear just like synchronous instruments, meant to be used inside the asynchronous callback. This proposal requires the callbacks to be allocated after the instruments, as follows:

```
asyncCounter := meter.NewAsyncCounter(...)
asyncGauge := meter.NewAsyncGauge(...)
cbfunc := func(ctx context.Context) {
    expensiveResult := expensiveCall()
    asyncCounter.Observe(ctx, expensiveResult.Count)
    asyncGauge.Observe(ctx, expensiveResult.Gauge)
}
meter.NewCallback(cbfunc, asyncCounter, asyncGauge)
```

This feels idiomatic to me. 

However the wording of the specification would not permit this. A slight adjustment can fix it.
Instead of saying that the instrument constructor accepts a callback, a separate statement, i.e., relax the spec to say the API MUST provide a way to associate instruments and callbacks.

### PR DESCRIPTION
**Pull request by jmacd**
Fixes #2280 

## Changes

This changes slightly the requirement for how instruments are associated with callbacks. Whereas before the spec required one instrument per callback, this would allow multiple instruments. A motivating example is given in #2280.

Related issues #2270 #2249

### CODE REVIEW

**Comment by jack-berg** [comment]
I believe we also need to loosen the requirement that APIs produce an error when multiple instruments are [registered with the same name](https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/metrics/api.md#L186-L187). 

Does that make sense to do in this PR or in a separate one?

**Comment by jmacd** [comment]
@jack-berg this PR definitely depends on #2270.

File: specification/metrics/api.md [hunk 1]
```diff
@@ -253,12 +253,13 @@ asynchronous:
...
 * Asynchronous instruments (e.g. [Asynchronous Gauge](#asynchronous-gauge)) give
-  the user a way to register callback function, and the callback function will
+  the user a way to register one or more associated callback functions, which will
   only be invoked upon collection. For example, a piece of embedded software
   could use an asynchronous gauge to collect the temperature from a sensor every
   15 seconds, which means the callback function will only be invoked every 15
-  seconds. [Measurements](#measurement) recorded by asynchronous instruments
-  cannot be associated with the [Context](../context/context.md).
+  seconds. Multiple callbacks may be associated with one instrument as a result
+  of duplicate instrument registration. [Measurements](#measurement) recorded by
```
Comment (Aneurysm9) [inline specification/metrics/api.md:261]: Does anything need to be specified regarding what should happen when multiple callbacks are associated with one instrument?

Comment (jmacd) [reply specification/metrics/api.md:261]: Maybe. I was hoping this is covered in #2270.

**Review by jsuereth — APPROVED** [review]
Thanks for tweaking this @jmacd !

**Review by jack-berg — APPROVED** [review]

**Comment by jack-berg** [comment]
It seems natural that if multiple callbacks allowed, that there should be some way of unregister / cancel / stop callbacks #2232. But I think that can be solved separately.

**Review by MrAlias — APPROVED** [review]

**Review by anuraaga — APPROVED** [review]

File: specification/metrics/api.md (hunk 1 above)
Comment (jmacd) [reply specification/metrics/api.md:261]: See #2226 for debate.

**Comment by jmacd** [comment]
Reviewers, as noted in https://github.com/open-telemetry/opentelemetry-specification/issues/2226#issuecomment-1028243274 I am closing this PR. This work will be integrated into #2270 following the discussion in #2226.

**Pull request closed without merging**